chalk/transform.py

```diff
@@ -47,123 +208,166 @@ def __rmul__(self, t: Affine) -> Self:
-            (
-                sa * oa + sb * od,
-                sa * ob + sb * oe,
-                sa * oc + sb * of + sc,
-                sd * oa + se * od,
-                sd * ob + se * oe,
-                sd * oc + se * of + sf,
-                0.0,
-                0.0,
-                1.0,
-            ),
-        )
-
-    elif hasattr(other, "from_points"):
-        # Point/vector array
-        points = getattr(other, "points", other)
-        try:
-            return other.from_points(
-                Point(px * sa + py * sb + sc, px * sd + py * se + sf)
-                for px, py in points
-            )
-        except TypeError:
-            return NotImplemented
+
+def ray_ray_intersection(
+    ray1: Tuple[P2_t, V2_t], ray2: Tuple[P2_t, V2_t]
+) -> Optional[Tuple[float, float]]:
+    """Given two rays
+
+    ray₁ = λ t . p₁ + t v₁
+    ray₂ = λ t . p₂ + t v₂
+
+    the function returns the parameters t₁ and t₂ at which the two rays meet,
+    that is:
+
+    ray₁ t₁ = ray₂ t₂
+
+    """
+    u = ray2[0] - ray1[0]
+    x1 = cross(ray1, ray2[1])
+    x2 = cross(u, ray1[1])
+    x3 = cross(u, ray2[1])
+    if x1 == 0 and x2 != 0:
+        # parallel
+        return None
+    elif x1 == 0 and x2 == 0:
+        return 0.0, 0.0
     else:
-        try:
-            vx, vy = other
-        except Exception:
-            return NotImplemented
-        return Vec2(vx * sa + vy * sb + sc, vx * sd + vy * se + sf)
+        # intersecting or collinear
+        return x3 / x1, x2 / x1
+
+
+def ray_circle_intersection(anchor: Float[Array, "... 3 1"], 
+                            direction: Float[Array, "... 3 1"], 
+                            circle_radius) -> Tuple[Float[Array, "..."],
+                                                    Int[Array, "..."]]:
+    """Given a ray and a circle centered at the origin, return the parameter t
+    where the ray meets the circle, that is:
+
+    ray t = circle θ
+
+    The above equation is solved as follows:
+
+    x + t v_x = r sin θ
+    y + t v_y = r cos θ
+
+    By squaring the equations and adding them we get
+
+    (x + t v_x)² + (y + t v_y)² = r²,
+
+    which is equivalent to the following equation:
+
+    (v_x² + v_y²) t² + 2 (x v_x + y v_y) t + (x² + y² - r²) = 0
+
+    This is a quadratic equation, whose solutions are well known.
+
+    """
+    a = length2(direction)
+    b = 2 * dot(anchor, direction)
+    c = length2(anchor) - circle_radius**2
+    Δ = b**2 - 4 * a * c
+    eps = 1e-12  # rounding error tolerance
+
+    mid = (((-eps <= Δ) & (Δ < eps)))[..., None]
+    mask = (Δ < -eps)[..., None] | (mid * np.array([1, 0]))
+
+    # Bump NaNs since they are going to me masked out. 
+    ret = np.stack([(-b - np.sqrt(Δ + 1e5 * mask[..., 0])) / (2 * a),
+                    (-b + np.sqrt(Δ + 1e5 * mask[..., 1])) / (2 * a)], -1)
+
+    ret = np.where(mid, (-b  / (2 * a))[..., None], ret)
+    return ret.transpose(2, 0, 1), 1- mask.transpose(2, 0, 1)
+
+
+    v = -b / (2 * a)
```

Comment:
The code here will not be reached because of the `return` statement above. But probably that's on purpose 🙂